### ISSUE TITLE
DllNotFoundException when building for Android 64-bit

### ISSUE THREAD

**Issue by weblings** [Funnyguy77/PocketSphinxUnityDemo#25]
I'm attempting to get an app approved on Oculus App Lab which requires a 64-bit application. After upgrading the build settings and the plugin settings from 32-bit I was getting these errors when PocketSphinx tried to initialize.

> 2021/02/04 23:44:29.257 15895 15921 Error Unity Unable to find libpocketsphinxwrap
> 2021/02/04 23:44:29.259 15895 15921 Error Unity Unable to find pocketsphinxwrap
> 2021/02/04 23:44:29.271 15895 15921 Error Unity DllNotFoundException: Unable to load DLL 'libpocketsphinxwrap': The specified module could not be found.
> 2021/02/04 23:44:29.271 15895 15921 Error Unity   at Pocketsphinx.pocketsphinxPINVOKE+SWIGExceptionHelper.SWIGRegisterExceptionCallbacks_pocketsphinx (Pocketsphinx.pocketsphinxPINVOKE+SWIGExceptionHelper+ExceptionDelegate applicationDelegate, Pocketsphinx.pocketsphinxPINVOKE+SWIGExceptionHelper+ExceptionDelegate arithmeticDelegate, Pocketsphinx.pocketsphinxPINVOKE+SWIGExceptionHelper+ExceptionDelegate divideByZeroDelegate, Pocketsphinx.pocketsphinxPINVOKE+SWIGExceptionHelper+ExceptionDelegate indexOutOfRangeDelegate, Pocketsphinx.pocketsphinxPINVOKE+SWIGExceptionHelper+ExceptionDelegate invalidCastDelegate, Pocketsphinx.pocketsphinxPINVOKE+SWIGExceptionHelper+ExceptionDelegate invalidOperationDelegate, Pocketsphinx.pocketsphinxPINVOKE+SWIGExceptionHelper+ExceptionDelegate ioDelegate, Pocketsphinx.pocketsphinxPINVOKE+SWIGExceptionHelper+ExceptionDelegate nullReferenceDelegate, Pocketsphinx.pocketsphinxPINVOKE+SWIGExceptionHelper+ExceptionDelegate outOfM

Making sure the .so is inside of the Plugins folder gets rid of the first few errors but the DllNotFoundException sticks around

> 2021/02/04 23:52:16.151 17296 17323 Error Unity DllNotFoundException: Unable to load DLL 'libpocketsphinxwrap': The specified module could not be found.
> 2021/02/04 23:52:16.151 17296 17323 Error Unity   at Pocketsphinx.pocketsphinxPINVOKE+SWIGExceptionHelper.SWIGRegisterExceptionCallbacks_pocketsphinx (Pocketsphinx.pocketsphinxPINVOKE+SWIGExceptionHelper+ExceptionDelegate applicationDelegate, Pocketsphinx.pocketsphinxPINVOKE+SWIGExceptionHelper+ExceptionDelegate arithmeticDelegate, Pocketsphinx.pocketsphinxPINVOKE+SWIGExceptionHelper+ExceptionDelegate divideByZeroDelegate, Pocketsphinx.pocketsphinxPINVOKE+SWIGExceptionHelper+ExceptionDelegate indexOutOfRangeDelegate, Pocketsphinx.pocketsphinxPINVOKE+SWIGExceptionHelper+ExceptionDelegate invalidCastDelegate, Pocketsphinx.pocketsphinxPINVOKE+SWIGExceptionHelper+ExceptionDelegate invalidOperationDelegate, Pocketsphinx.pocketsphinxPINVOKE+SWIGExceptionHelper+ExceptionDelegate ioDelegate, Pocketsphinx.pocketsphinxPINVOKE+SWIGExceptionHelper+ExceptionDelegate nullReferenceDelegate, Pocketsphinx.pocketsphinxPINVOKE+SWIGExceptionHelper+ExceptionDelegate outOfM

'Load on Startup' does not affect this behavior at all. This is how I have the plugin setup.
![image](https://user-images.githubusercontent.com/8001710/107005926-4603f180-6745-11eb-9f5d-4786ce3bb580.png)

Here's my project's target architecture settings.
![image](https://user-images.githubusercontent.com/8001710/107004547-4d2a0000-6743-11eb-95db-c08b478b3676.png)

Hopefully the error is on my end and there is a simple solution, but thought I'd submit an issue just in case.

**Comment by ghost** [comment]
I've encountered the same problem, especially because of Google Play Store changes that require separate support for ARM64.

The problem is because the library you are talking about is compiled for ARMv7, with Scripting Backend set to Mono, not to IL2CPP. However, Mono provides support only for 32 bits, and even if 64-bits-based phones have no problem to install, open and use it - Google doesn't accept that.

Question from my site to the author: would you like to provide .so library for 64-bit / ARM64 Android?

**Comment by Funnyguy77** [comment]
You're correct, this package doesn't include an ARM64 library. However, I have been able to resolve this issue in the past. 

This [Link](https://github.com/NielsMasdorp/Speculum-Android/blob/master/app/src/main/jniLibs/arm64-v8a/libpocketsphinx_jni.so) provides a working ARM64 library. However, I wasn't able to add it in Unity directly but I was able to export the project to android studio and add the missing library there before generating a build.

That should work for you. 👍

**Comment by ghost** [comment]
Ok, thanks. I've tried to just replace that file in my Unity project. It's loaded in runtime properly (I don't get any error with loading library using IL2CPP/ARM64 stuff), but probably it's written different and there're other methods. I will try to handle it with this what I have, maybe example implementation is enough to modify c# scripts.

**Comment by weblings** [comment]
Thanks for the lead and response @Funnyguy77 ! 

I did what you suggested and exported from Unity to Android Studio. I renamed that .so file to match the name of the .so from this repo and added it in the "arm64-v8a" folder. Running the build, it is able to find the file now (yay!), but now gets stuck looking for "SWIGRegisterExceptionCallbacks_pocketsphinx" Here's what I'm getting out of adb:

>  <color=red>Initializing decoder...</color>
> 03-14 18:43:29.353 28526 28542 I Unity   : NoDetection_PocketSphinx:SetupDecoderKWS()
> 03-14 18:43:29.353 28526 28542 I Unity   : <Start>d__11:MoveNext()
> 03-14 18:43:29.353 28526 28542 I Unity   : UnityEngine.SetupCoroutine:InvokeMoveNext(IEnumerator, IntPtr)
> 03-14 18:43:29.353 28526 28542 I Unity   :
> 03-14 18:43:29.353 28526 28542 I Unity   : (Filename: ./Runtime/Export/Debug/Debug.bindings.h Line: 35)
> 03-14 18:43:29.353 28526 28542 I Unity   :
> 03-14 18:43:29.359 28526 28542 E Unity   : EntryPointNotFoundException: Unable to find an entry point named 'SWIGRegisterExceptionCallbacks_pocketsphinx' in 'libpocketsphinxwrap'.
> 03-14 18:43:29.359 28526 28542 E Unity   :   at Pocketsphinx.pocketsphinxPINVOKE+SWIGExceptionHelper.SWIGRegisterExceptionCallbacks_pocketsphinx (Pocketsphinx.pocketsphinxPINVOKE+SWIGExceptionHelper+ExceptionDelegate applicationDelegate, Pocketsphinx.pocketsphinxPINVOKE+SWIGExceptionHelper+ExceptionDelegate arithmeticDelegate, Pocketsphinx.pocketsphinxPINVOKE+SWIGExceptionHelper+ExceptionDelegate divideByZeroDelegate, Pocketsphinx.pocketsphinxPINVOKE+SWIGExceptionHelper+ExceptionDelegate indexOutOfRangeDelegate, Pocketsphinx.pocketsphinxPINVOKE+SWIGExceptionHelper+ExceptionDelegate invalidCastDelegate, Pocketsphinx.pocketsphinxPINVOKE+SWIGExceptionHelper+ExceptionDelegate invalidOperationDelegate, Pocketsphinx.pocketsphinxPINVOKE+SWIGExceptionHelper+ExceptionDelegate ioDelegate, Pocketsphinx.pocketsphinxPINVOKE+SWIGExceptionHelper+ExceptionDelegate nullReferenceDelegate, Pocketsphinx.pocketsphinxPINVOKE+SWIGExceptio

I don't think there's much that can be done since neither of us compiled this .so file, but maybe you have an idea?

The script that is running is [SphinxExample.cs](https://github.com/Funnyguy77/PocketSphinxUnityDemo/blob/4adf9f70d6f7e84067b2a5b637d2f5ac07904e85/Assets/Scripts/SphinxExample.cs)

**Comment by SubatomicPlanets** [comment]
Hello there. I Also have the same issue and am getting an error with "SWIGRegisterExceptionCallbacks_pocketsphinx" not being found. I don't know much about .so or .dll files and I have no clue what to do now. It would be brilliant if someone can help me or at least tell me how to build the .so file myself. If I find something to make it work I will let you know. Thanks!
